src/main/java/apoc/concurrent/Semaphores.java

```diff
@@ -0,0 +1,116 @@
+package apoc.concurrent;
+
+import apoc.ApocConfig;
+import org.neo4j.internal.helpers.collection.Iterators;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.internal.LogService;
+import org.neo4j.procedure.TerminationGuard;
+import org.neo4j.procedure.impl.GlobalProceduresRegistry;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * holds a per-database level map of declared semaphores
+ */
+public class Semaphores extends LifecycleAdapter {
+
+    private final Map<String, Semaphore> semaphoreMap = new ConcurrentHashMap<>();
+
+    private final Log log;
+    private final GlobalProceduresRegistry globalProceduresRegistry;
+    private final ApocConfig apocConfig;
+
+    public Semaphores(LogService log, GlobalProceduresRegistry globalProceduresRegistry, ApocConfig apocConfig) {
+
+        this.log = log.getInternalLog(Semaphores.class);
+        this.globalProceduresRegistry = globalProceduresRegistry;
+        this.apocConfig = apocConfig;
+
+        // expose this config instance via `@Context ApocConfig config`
+        globalProceduresRegistry.registerComponent((Class<Semaphores>) getClass(), ctx -> this, true);
```

Comment:
we should probably also register other components that we currently access with static